src/helpers/fabric.rs

```diff
@@ -0,0 +1,56 @@
+use crate::helpers::error::Error;
+use crate::helpers::Identity;
+use crate::protocol::{RecordId, Step};
+use async_trait::async_trait;
+use futures::Stream;
+use std::fmt::{Debug, Formatter};
+
+/// Combination of helper identity and step that uniquely identifies a single channel of communication
+/// between two helpers.
+#[derive(Copy, Clone, Eq, PartialEq, Hash)]
+pub struct ChannelId<S> {
+    pub identity: Identity,
+    pub step: S,
+}
+
+#[derive(Debug)]
+pub struct MessageEnvelope {
+    pub record_id: RecordId,
+    pub payload: Box<[u8]>,
```

Comment:
You mentioned `SmallVec` as providing advantage here.  Another TODO?

Reply:
yes, it is on my todo list :). I did not want to modify it in this PR to make sure I can accurately measure the impact of it and to make this change easier to review.

I was planning to start working on it after this change - it is more critical that we have `Fabric` interface ready so @thurstonsand can start making progress on integrating networking layer

![Screen Shot 2022-10-11 at 12 02 22 PM](https://user-images.githubusercontent.com/1110516/195177047-1c45dc2e-6723-4495-83f7-888b83009d4b.png)